cpp/src/arrow/dataset/scanner.cc

```diff
@@ -877,6 +882,15 @@ Result<compute::ExecNode*> MakeScanNode(compute::ExecPlan* plan,
         batch->values.emplace_back(partial.fragment.index);
         batch->values.emplace_back(partial.record_batch.index);
         batch->values.emplace_back(partial.record_batch.last);
+
+        auto filefragment_casted = dynamic_cast<const FileFragment*>(
+            partial.fragment.value.get());
+        if (filefragment_casted != nullptr) {
+          batch->values.emplace_back(filefragment_casted->ToString());
+        } else {
+          batch->values.emplace_back("Not a file fragment");
```

Comment:
The `ToString()` method will return the path if it is a `FileFragment` or will return the `type_name` otherwise, so we may ignore doing the cast and directly emplace back the string returned from `ToString()`